### DISCUSSION TITLE
New functional ! Consider, please, to add it !

### DISCUSSION THREAD

**Discussion by CpServiceSpb** [mvp/uhubctl#546] [General]
UHubCtl shows and manage compatible deices and its ports in/by non common human format, for example (RSHTech A13) , per device - means per each physical device:
```
Current status for hub 2-5.4 [0bda:5411 Generic USB2.1 Hub, USB 2.10, 4 ports, ppps]
  Port 1: 0503 power highspeed enable connect Device ...
  Port 2: 0503 power highspeed enable connect Device ...
  Port 3: 0503 power highspeed enable connect Device ...
  Port 4: 0503 power highspeed enable connect Device ...
Current status for hub 2-5.3.4 [0bda:5411 Generic USB2.1 Hub, USB 2.10, 4 ports, ppps]
  Port 1: 0100 power
  Port 2: 0503 power highspeed enable connect Device ...
  Port 3: 0103 power enable connect Device ...
  Port 4: 0103 power 
Current status for hub 2-5.3 [0bda:5411 Generic USB2.1 Hub, USB 2.10, 4 ports, ppps]
  Port 1: 0503 power highspeed enable connect Device ...
  Port 2: 0503 power 
  Port 3: 0103 power enable connect Device ...
  Port 4: 0503 power highspeed enable connect [0bda:5411 Generic USB2.1 Hub, USB 2.10, 4 ports, ppps]
Current status for hub 2-5 [0bda:5411 Generic USB2.1 Hub, USB 2.10, 4 ports, ppps]
  Port 1: 0100 power
  Port 2: 0503 power highspeed enable connect Device ...
  Port 3: 0503 power highspeed enable connect [0bda:5411 Generic USB2.1 Hub, USB 2.10, 4 ports, ppps]
  Port 4: 0503 power highspeed enable connect [0bda:5411 Generic USB2.1 Hub, USB 2.10, 4 ports, ppps]
```
I recommend to add the parameter, if it is specified, output will be in such way, but to make human well readable output, such for single physical device:
```
 Port 1: power highspeed enable connect Device ...
 Port 2: power enable connect Device ...
 Port 3: power highspeed enable connect Device ...
.................
Port 11: power enable connect Device ...
Port 12: power enable connect Device ...
Port 13: power highspeed enable connect Device ...
```
Here is port by order number and port by logical device mapping from above example:
```
[Port 1 is port 1 of the 1st "subhub" = Port 1 of 2-5 logical device;
Port 2 is port 2 of the 1st "subhub" = Port 2 of 2-5 logical device;
.....
Port 5 is port 1 of the 2nd "subhub" = Port 1 of 2-5.3 logical device;
Port 6 is port 2 of the 2nd "subhub" = Port 2 of 2-5.3 logical device;
...
Port 9   is port 1 of the 3nd "subhub" = Port 1 of 2-5.3.4 logical device;
Port 10 is port 2 of the 3nd "subhub" = Port 2 of 2-5.3.4 logical device;
...
Port 13 is port 1 of the 4th "subhub" = Port 1 of 2-5.4 logical device;
Port 14 is port 2 of the 4th"subhub" = Port 2 of 2-5.4 logical device;
...
Port 16 is port 4 of the 4th"subhub" = Port 4 of 2-5.4 logical device;
```
Why  is totally 16 ports for A13 hub - because of the interhub connections.
B the way these "ports" is able to exclude from the list.
And it will be well to manage of USB devices by port order number.

As such hubs doesn' t have feature to store all USB ports state after reboot host computer or power off/on itself, it would be better to add some soft saving ports state - may be add demonize part - during restart/shutdown/poweroff of a host OS, daemon get appropriate signal, read specified at it settings file hub or hubs and save to small DB (even txt) state of each port of each hub - for optimization it is enough to save only off ports state, bacause anyway after rebooting host computer/ power off/on hubs all port will be on by defaul.
And when OS starts, daemon part starts, read previously saved ports state of specified hub, where there are only off ports and move all these ports to off state.

Please take a think about it two features and it would better that you will implement it.

**Comment by mvp** [comment]
Unfortunately, this would require creating and maintaining custom port map for each USB hub in existence:
* Maintaining this map is expensive, as it will have to be present and compiled into uhubctl source code, and captured for all hubs up front
* Often, manufacturers release devices with exact same model names and part numbers, yet with different internal topology. E.g. some DLink DUB-H7 have one true 7 port chip, and some have 2x4 port chips
* It is not that difficult for you, as hub owner, to figure out internal mapping and attach stickers near hub ports that you can use to drive it correctly.

Long story short, sorry, but this is not going to happen.
